src/machine/image.c

```diff
@@ -0,0 +1,207 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include "sd-json.h"
+
+#include "fd-util.h"
+#include "fileio.h"
+#include "discover-image.h"
+#include "image.h"
+#include "io-util.h"
+#include "operation.h"
+#include "process-util.h"
+#include "string-table.h"
+#include "tmpfile-util.h"
+
+int clean_pool_read_first_entry(FILE *file, int child_error, sd_bus_error *error) {
+        _cleanup_free_ char *name = NULL;
+        bool success;
+        int r;
+
+        assert(file);
+
+        if (fseek(file, 0, SEEK_SET) < 0)
+                return log_debug_errno(errno, "Failed to seek to the beggining of tmp file: %m");
+
+        /* The resulting temporary file starts with a boolean value that indicates success or not. */
+        errno = 0;
+        size_t n = fread(&success, 1, sizeof(success), file);
+        if (n != sizeof(success)) {
+                log_debug_errno(errno_or_else(EIO), "Received unexpected amount of bytes: %m");
+                return child_error < 0 ? child_error : errno_or_else(EIO);
+        }
+
+        if (child_error < 0) {
+                /* The clean-up operation failed. In this case the resulting temporary file should contain a boolean
+                 * set to false followed by the name of the failed image. Let's try to read this and use it for the
+                 * error message. If we can't read it, don't mind, and return the naked error. */
+
+                if (success) { /* The resulting temporary file could not be updated, ignore it. */
+                        log_debug("Child process failed but failed to write name of failed image to tmp file");
+                        return child_error;
+                }
+
+                r = read_nul_string(file, LONG_LINE_MAX, &name);
+                if (r <= 0) { /* Same here... */
+                        log_debug_errno(r, "Failed to read NULL-terminated string from tmp file: %m");
+                        return child_error;
+                }
+
+                log_debug_errno(child_error, "Failed to remove image '%s': %m", name);
+                return sd_bus_error_set_errnof(error, child_error, "Failed to remove image %s: %m", name);
+        }
+
+        if (!success)
+                return log_debug_errno(SYNTHETIC_ERRNO(EINVAL), "Inconsistent state of the temporary file");
```

Comment:
`EINVAL` is for invalid argument or so. Maybe `ESTALE` or so is better?

Also, period please.